File: app/src/main/java/cf/playhi/freezeyou/ui/BackupImportChooserActivity.java
package cf.playhi.freezeyou.ui;

import android.content.Intent;
import android.os.Bundle;
import android.util.Base64;
import android.widget.Button;
import android.widget.ListView;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;

import cf.playhi.freezeyou.R;
import cf.playhi.freezeyou.adapter.BackupImportChooserActivitySwitchSimpleAdapter;
import cf.playhi.freezeyou.app.FreezeYouBaseActivity;
import cf.playhi.freezeyou.storage.key.AbstractKey;
import cf.playhi.freezeyou.storage.key.DefaultMultiProcessMMKVStorageBooleanKeys;
import cf.playhi.freezeyou.storage.key.DefaultMultiProcessMMKVStorageStringKeys;
import cf.playhi.freezeyou.storage.key.DefaultSharedPreferenceStorageBooleanKeys;
import cf.playhi.freezeyou.storage.key.DefaultSharedPreferenceStorageStringKeys;
import cf.playhi.freezeyou.utils.BackupUtils;
import cf.playhi.freezeyou.utils.ToastUtils;

import static cf.playhi.freezeyou.utils.ThemeUtils.processActionBar;
import static cf.playhi.freezeyou.utils.ThemeUtils.processSetTheme;

public class BackupImportChooserActivity extends FreezeYouBaseActivity {

    private final HashMap<String, String> keyToStringIdValuePair = new HashMap<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        processSetTheme(this);
        super.onCreate(savedInstanceState);
        processActionBar(getSupportActionBar());
        setContentView(R.layout.bica_main);

        onCreateInit();
    }

    private void onCreateInit() {

        Intent intent = getIntent();
        if (intent == null) {
            finish();
            return;
        }

        final ListView mainListView = findViewById(R.id.bica_main_listView);
        final ArrayList<HashMap<String, String>> titleAndSpKeyArrayList = new ArrayList<>();

        generateKeyToStringIdValuePair();

        String jsonContentString = intent.getStringExtra("jsonObjectString");
        JSONObject jsonObject = null;
        if (jsonContentString == null) {
            HashMap<String, String> keyValuePair = new HashMap<>();
            keyValuePair.put("title", getString(R.string.failed));
            keyValuePair.put("spKey", "Failed!");
            keyValuePair.put("category", "Failed!");
            titleAndSpKeyArrayList.add(keyValuePair);
        } else {
            try {
                jsonObject = new JSONObject(jsonContentString);
            } catch (JSONException e) {
                e.printStackTrace();
            }
            if (jsonObject == null) {
                HashMap<String, String> keyValuePair = new HashMap<>();
                keyValuePair.put("title", getString(R.string.parseFailed));
                keyValuePair.put("spKey", "Failed!");
                keyValuePair.put("category", "Failed!");
                titleAndSpKeyArrayList.add(keyValuePair);
            } else {
                generateList(jsonObject, titleAndSpKeyArrayList);
                if (titleAndSpKeyArrayList.size() == 0) {
                    HashMap<String, String> keyValuePair = new HashMap<>();
                    keyValuePair.put("title", getString(R.string.nothing));
                    keyValuePair.put("spKey", "Failed!");
                    keyValuePair.put("category", "Failed!");
                    titleAndSpKeyArrayList.add(keyValuePair);
                }
            }
        }

        final BackupImportChooserActivitySwitchSimpleAdapter adapter =
                new BackupImportChooserActivitySwitchSimpleAdapter(
                        this,
                        jsonObject,
                        titleAndSpKeyArrayList,
                        R.layout.bica_list_item,
                        new String[]{"title"},
                        new int[]{R.id.bica_list_item_switch});

        mainListView.setAdapter(adapter);

        processButtons();
    }

    private void processButtons() {
        final Button bicaFinishButton = findViewById(R.id.bica_finish_button);
        final Button bicaCancelButton = findViewById(R.id.bica_cancel_button);
        bicaCancelButton.setOnClickListener(v -> finish());
        bicaFinishButton.setOnClickListener(v -> {
            final ListView mainListView = findViewById(R.id.bica_main_listView);
            BackupImportChooserActivitySwitchSimpleAdapter adapter =
                    (BackupImportChooserActivitySwitchSimpleAdapter) mainListView.getAdapter();
            BackupUtils.importContents(
                    getApplicationContext(),
                    BackupImportChooserActivity.this, adapter.getFinalList());
            ToastUtils.showToast(BackupImportChooserActivity.this, R.string.finish);
            finish();
        });
    }

    private void generateKeyToStringIdValuePair() {
        // Int 开始
        keyToStringIdValuePair.put("onClickFunctionStatus", getString(R.string.onClickFunctionStatus));
        keyToStringIdValuePair.put("sortMethodStatus", getString(R.string.sortMethodStatus));
        // Int 结束
        // 一键冻结、一键解冻、离开冻结列表 开始
        keyToStringIdValuePair.put("okff", getString(R.string.oneKeyFreezeList));
        keyToStringIdValuePair.put("okuf", getString(R.string.oneKeyUFList));
        keyToStringIdValuePair.put("foq", getString(R.string.freezeOnceQuitList));
        // 一键冻结、一键解冻、离开冻结列表 结束
        // 安装应用请求、URI 请求白名单 开始
        keyToStringIdValuePair.put("uriAutoAllowPkgs_allows", getString(R.string.uriAutoAllowList));
        keyToStringIdValuePair.put("installPkgs_autoAllowPkgs_allows", getString(R.string.ipaAutoAllowList));
        // 安装应用请求、URI 请求白名单 结束
    }

    private void generateList(JSONObject jsonObject, ArrayList<HashMap<String, String>> list) {

        final Iterator<String> jsonKeysIterator = jsonObject.keys();
        while (jsonKeysIterator.hasNext()) {
            switch (jsonKeysIterator.next()) {
                // 通用设置转入（更多设置 中的选项，不转移图标选择相关设置） 开始
                case "generalSettings_boolean":
                    generateGeneralSettingsBooleanList(jsonObject, list);
                    break;
                case "generalSettings_string":
                    generateGeneralSettingsStringList(jsonObject, list);
                    break;
                case "generalSettings_int":
                    generateGeneralSettingsIntList(jsonObject, list);
                    break;
                // 通用设置转出 结束
                // 一键冻结、一键解冻、离开冻结列表 开始
                case "oneKeyList":
                    generateOneKeyList(jsonObject, list);
                    break;
                // 一键冻结、一键解冻、离开冻结列表 结束
                // 计划任务 - 时间 开始
                case "userTimeScheduledTasks":
                    generateUserTimeScheduledTasksList(jsonObject, list);
                    break;
                // 计划任务 - 时间 结束
                // 计划任务 - 触发器 开始
                case "userTriggerScheduledTasks":
                    generateUserTriggerScheduledTasksList(jsonObject, list);
                    break;
                // 计划任务 - 触发器 结束
                // 用户自定分类（我的列表） 开始
                case "userDefinedCategories":
                    generateUserDefinedCategoriesList(jsonObject, list);
                    break;
                // 用户自定分类（我的列表） 结束
                // URI 请求白名单 开始
                case "uriAutoAllowPkgs_allows":
                    generateUriAutoAllowPkgsList(jsonObject, list);
                    break;
                // URI 请求白名单 结束
                // 安装应用请求白名单 开始
                case "installPkgs_autoAllowPkgs_allows":
                    generateInstallPkgsAutoAllowPkgsList(jsonObject, list);
                    break;
                // 安装应用请求白名单 结束
                default:
                    break;
            }
        }
    }

    private void generateGeneralSettingsBooleanList(JSONObject jsonObject, ArrayList<HashMap<String, String>> list) {
        JSONArray array = jsonObject.optJSONArray("generalSettings_boolean");
        if (array == null) {
            return;
        }
        JSONObject generalSettingsBooleanJSONObject = array.optJSONObject(0);
        if (generalSettingsBooleanJSONObject == null) {
            return;
        }
        Iterator<String> it = generalSettingsBooleanJSONObject.keys();
        String moreSettingsDashLineLabel = getString(R.string.moreSettingsDashLineLabel);
        String s;
        while (it.hasNext()) {
            s = it.next();

            if (s == null) continue;

            if (DefaultSharedPreferenceStorageBooleanKeys.firstIconEnabled.name().equals(s)
                    || DefaultSharedPreferenceStorageBooleanKeys.secondIconEnabled.name().equals(s)
                    || DefaultSharedPreferenceStorageBooleanKeys.thirdIconEnabled.name().equals(s)
                    || DefaultMultiProcessMMKVStorageBooleanKeys.enableAuthentication.name().equals(s)) {
                continue;
            }

            AbstractKey<Boolean> key = null;

            try {
                key = DefaultSharedPreferenceStorageBooleanKeys.valueOf(s);
            } catch (IllegalArgumentException ignored) {
            }

            if (key == null) {
                try {
                    key = DefaultMultiProcessMMKVStorageBooleanKeys.valueOf(s);
                } catch (IllegalArgumentException ignored) {
                }
            }

            if (key == null) continue;

            HashMap<String, String> keyValuePair = new HashMap<>();
            keyValuePair.put(
                    "title",
                    String.format(moreSettingsDashLineLabel, getString(key.titleTextStringId()))
            );
            keyValuePair.put("spKey", s);
            keyValuePair.put("category", "generalSettings_boolean");
            list.add(keyValuePair);
        }
    }

    private void generateGeneralSettingsIntList(JSONObject jsonObject, ArrayList<HashMap<String, String>> list) {
        JSONArray array = jsonObject.optJSONArray("generalSettings_int");
        if (array == null) {
            return;
        }
        JSONObject generalSettingsIntJSONObject = array.optJSONObject(0);
        if (generalSettingsIntJSONObject == null) {
            return;
        }
        Iterator<String> it = generalSettingsIntJSONObject.keys();
        String s;
        while (it.hasNext()) {
            s = it.next();
            switch (s) {
                case "onClickFunctionStatus":
                case "sortMethodStatus":
                    HashMap<String, String> keyValuePair = new HashMap<>();
                    keyValuePair.put("title",
                            keyToStringIdValuePair.containsKey(s) ?
                                    keyToStringIdValuePair.get(s) : s);
                    keyValuePair.put("spKey", s);
                    keyValuePair.put("category", "generalSettings_int");
                    list.add(keyValuePair);
                    break;
                default:
                    break;
            }
        }
    }

    private void generateGeneralSettingsStringList(JSONObject jsonObject, ArrayList<HashMap<String, String>> list) {
        JSONArray array = jsonObject.optJSONArray("generalSettings_string");
        if (array == null) {
            return;
        }
        JSONObject generalSettingsStringJSONObject = array.optJSONObject(0);
        if (generalSettingsStringJSONObject == null) {
            return;
        }
        String moreSettingsDashLineLabel = getString(R.string.moreSettingsDashLineLabel);
        Iterator<String> it = generalSettingsStringJSONObject.keys();
        String s;
        while (it.hasNext()) {
            s = it.next();

            if (s == null) continue;

            AbstractKey<String> key = null;

            try {
                key = DefaultSharedPreferenceStorageStringKeys.valueOf(s);
            } catch (IllegalArgumentException ignored) {
            }

            if (key == null) {
                try {
                    key = DefaultMultiProcessMMKVStorageStringKeys.valueOf(s);
                } catch (IllegalArgumentException ignored) {
                }
            }

            if (key == null) continue;

            HashMap<String, String> keyValuePair = new HashMap<>();
            keyValuePair.put(
                    "title",
                    String.format(moreSettingsDashLineLabel, getString(key.titleTextStringId()))
            );
            keyValuePair.put("spKey", s);
            keyValuePair.put("category", "generalSettings_string");
            list.add(keyValuePair);
        }
    }

    private void generateOneKeyList(JSONObject jsonObject, ArrayList<HashMap<String, String>> list) {
        JSONArray array = jsonObject.optJSONArray("oneKeyList");
        if (array == null) {
            return;
        }
        JSONObject generalOneKeyJSONObject = array.optJSONObject(0);
        if (generalOneKeyJSONObject == null) {
            return;
        }
        Iterator<String> it = generalOneKeyJSONObject.keys();
        String s;
        while (it.hasNext()) {
            s = it.next();
            switch (s) {
                case "okff":
                case "okuf":
                case "foq":
                    HashMap<String, String> keyValuePair = new HashMap<>();
                    keyValuePair.put("title",
                            keyToStringIdValuePair.containsKey(s) ?
                                    keyToStringIdValuePair.get(s) : s);
                    keyValuePair.put("spKey", s);
                    keyValuePair.put("category", "oneKeyList");
                    list.add(keyValuePair);
                    break;
                default:
                    break;
            }
        }
    }

    private void generateUserTimeScheduledTasksList(JSONObject jsonObject, ArrayList<HashMap<String, String>> list) {
        JSONArray array = jsonObject.optJSONArray("userTimeScheduledTasks");
        if (array == null) {
            return;
        }

        JSONObject oneUserTimeScheduledTaskJSONObject;
        String scheduledTaskDashLineLabel = getString(R.string.scheduledTaskDashLineLabel);
        String defaultLabel = getString(R.string.label);
        for (int i = 0; i < array.length(); ++i) {
            oneUserTimeScheduledTaskJSONObject = array.optJSONObject(i);
            if (oneUserTimeScheduledTaskJSONObject == null) {
                continue;
            }
            HashMap<String, String> keyValuePair = new HashMap<>();
            keyValuePair.put("title",
                    String.format(scheduledTaskDashLineLabel,
                            oneUserTimeScheduledTaskJSONObject.optString("label", defaultLabel)));
            try {
                oneUserTimeScheduledTaskJSONObject.put("i", Integer.toString(i));
            } catch (JSONException e) {
                e.printStackTrace();
            }
            keyValuePair.put("spKey", Integer.toString(i));
            keyValuePair.put("category", "userTimeScheduledTasks");
            list.add(keyValuePair);
        }
    }

    private void generateUserTriggerScheduledTasksList(JSONObject jsonObject, ArrayList<HashMap<String, String>> list) {
        JSONArray array = jsonObject.optJSONArray("userTriggerScheduledTasks");
        if (array == null) {
            return;
        }

        JSONObject oneUserTriggerScheduledTaskJSONObject;
        String scheduledTaskDashLineLabel = getString(R.string.scheduledTaskDashLineLabel);
        String defaultLabel = getString(R.string.label);
        for (int i = 0; i < array.length(); ++i) {
            oneUserTriggerScheduledTaskJSONObject = array.optJSONObject(i);
            if (oneUserTriggerScheduledTaskJSONObject == null) {
                continue;
            }
            HashMap<String, String> keyValuePair = new HashMap<>();
            keyValuePair.put("title",
                    String.format(scheduledTaskDashLineLabel,
                            oneUserTriggerScheduledTaskJSONObject.optString("label", defaultLabel)));
            try {
                oneUserTriggerScheduledTaskJSONObject.put("i", Integer.toString(i));
            } catch (JSONException e) {
                e.printStackTrace();
            }
            keyValuePair.put("spKey", Integer.toString(i));
            keyValuePair.put("category", "userTriggerScheduledTasks");
            list.add(keyValuePair);
        }
    }

    private void generateUserDefinedCategoriesList(JSONObject jsonObject, ArrayList<HashMap<String, String>> list) {
        JSONArray array = jsonObject.optJSONArray("userDefinedCategories");
        if (array == null) {
            return;
        }

        JSONObject oneUserDefinedCategoriesJSONObject;
        String myCustomizationDashLineLabel = getString(R.string.myCustomizationDashLineLabel);
        String defaultLabel = "";
        for (int i = 0; i < array.length(); ++i) {
            oneUserDefinedCategoriesJSONObject = array.optJSONObject(i);
            if (oneUserDefinedCategoriesJSONObject == null) {
                continue;
            }
            HashMap<String, String> keyValuePair = new HashMap<>();
            keyValuePair.put("title",
                    String.format(myCustomizationDashLineLabel,
                            new String(
                                    Base64.decode(
                                            oneUserDefinedCategoriesJSONObject.optString(
                                                    "label", defaultLabel),
                                            Base64.DEFAULT
                                    )
                            )
                    )
            );
            try {
                oneUserDefinedCategoriesJSONObject.put("i", Integer.toString(i));
            } catch (JSONException e) {
                e.printStackTrace();
            }
            keyValuePair.put("spKey", Integer.toString(i));
            keyValuePair.put("category", "userDefinedCategories");
            list.add(keyValuePair);
        }
    }

    private void generateUriAutoAllowPkgsList(JSONObject jsonObject, ArrayList<HashMap<String, String>> list) {
        JSONArray array = jsonObject.optJSONArray("uriAutoAllowPkgs_allows");
        if (array == null) {
            return;
        }

        JSONObject jObj = array.optJSONObject(0);
        if (jObj == null) {
            return;
        }

        HashMap<String, String> keyValuePair = new HashMap<>();
        keyValuePair.put(
                "title",
                keyToStringIdValuePair.containsKey("uriAutoAllowPkgs_allows") ?
                        keyToStringIdValuePair.get("uriAutoAllowPkgs_allows") :
                        "uriAutoAllowPkgs_allows"
        );
        keyValuePair.put("spKey", "uriAutoAllowPkgs_allows");
        keyValuePair.put("category", "uriAutoAllowPkgs_allows");
        list.add(keyValuePair);
    }

    private void generateInstallPkgsAutoAllowPkgsList(JSONObject jsonObject, ArrayList<HashMap<String, String>> list) {
        JSONArray array = jsonObject.optJSONArray("installPkgs_autoAllowPkgs_allows");
        if (array == null) {
            return;
        }

        JSONObject jObj = array.optJSONObject(0);
        if (jObj == null) {
            return;
        }

        HashMap<String, String> keyValuePair = new HashMap<>();
        keyValuePair.put(
                "title",
                keyToStringIdValuePair.containsKey("installPkgs_autoAllowPkgs_allows") ?
                        keyToStringIdValuePair.get("installPkgs_autoAllowPkgs_allows") :
                        "installPkgs_autoAllowPkgs_allows"
        );
        keyValuePair.put("spKey", "installPkgs_autoAllowPkgs_allows");
        keyValuePair.put("category", "installPkgs_autoAllowPkgs_allows");
        list.add(keyValuePair);
    }

}

File: app/src/main/java/cf/playhi/freezeyou/ui/BackupImportChooserActivity.kt
package cf.playhi.freezeyou.ui

import android.os.Bundle
import android.util.Base64
import androidx.activity.compose.setContent
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.itemsIndexed
import androidx.compose.material3.Button
import androidx.compose.material3.Switch
import androidx.compose.material3.Text
import androidx.compose.runtime.mutableStateMapOf
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import cf.playhi.freezeyou.R
import cf.playhi.freezeyou.app.FreezeYouBaseActivity
import cf.playhi.freezeyou.storage.key.DefaultMultiProcessMMKVStorageBooleanKeys
import cf.playhi.freezeyou.storage.key.DefaultMultiProcessMMKVStorageStringKeys
import cf.playhi.freezeyou.storage.key.DefaultSharedPreferenceStorageBooleanKeys
import cf.playhi.freezeyou.storage.key.DefaultSharedPreferenceStorageStringKeys
import cf.playhi.freezeyou.utils.BackupUtils
import cf.playhi.freezeyou.utils.ThemeUtils.processActionBar
import cf.playhi.freezeyou.utils.ThemeUtils.processSetTheme
import cf.playhi.freezeyou.utils.ToastUtils
import cf.playhi.freezeyou.ui.compose.FreezeYouTheme
import org.json.JSONException
import org.json.JSONObject

class BackupImportChooserActivity : FreezeYouBaseActivity() {
    private val keyToStringIdValuePair = HashMap<String, String>()
    private val selectedItems = mutableStateMapOf<Int, Boolean>()
    private var importObject: JSONObject? = null
    private var importItems: List<MutableMap<String, String>> = emptyList()

    override fun onCreate(savedInstanceState: Bundle?) {
        processSetTheme(this)
        super.onCreate(savedInstanceState)
        processActionBar(supportActionBar)
        onCreateInit()
    }

    private fun onCreateInit() {
        val intent = intent
        if (intent == null) {
            finish()
            return
        }
        val titleAndSpKeyArrayList = ArrayList<MutableMap<String, String>>()
        generateKeyToStringIdValuePair()
        val jsonContentString = intent.getStringExtra("jsonObjectString")
        var jsonObject: JSONObject? = null
        if (jsonContentString == null) {
            val keyValuePair = HashMap<String, String>()
            keyValuePair["title"] = getString(R.string.failed)
            keyValuePair["spKey"] = "Failed!"
            keyValuePair["category"] = "Failed!"
            titleAndSpKeyArrayList.add(keyValuePair)
        } else {
            try {
                jsonObject = JSONObject(jsonContentString)
            } catch (e: JSONException) {
                e.printStackTrace()
            }
            if (jsonObject == null) {
                val keyValuePair = HashMap<String, String>()
                keyValuePair["title"] = getString(R.string.parseFailed)
                keyValuePair["spKey"] = "Failed!"
                keyValuePair["category"] = "Failed!"
                titleAndSpKeyArrayList.add(keyValuePair)
            } else {
                generateList(jsonObject, titleAndSpKeyArrayList)
                if (titleAndSpKeyArrayList.size == 0) {
                    val keyValuePair = HashMap<String, String>()
                    keyValuePair["title"] = getString(R.string.nothing)
                    keyValuePair["spKey"] = "Failed!"
                    keyValuePair["category"] = "Failed!"
                    titleAndSpKeyArrayList.add(keyValuePair)
                }
            }
        }
        importObject = jsonObject?.let { JSONObject(it.toString()) }
        importItems = titleAndSpKeyArrayList
        titleAndSpKeyArrayList.indices.forEach { selectedItems[it] = true }
        setContent {
            FreezeYouTheme {
                Column(Modifier.fillMaxSize()) {
                    LazyColumn(Modifier.fillMaxWidth().weight(1f)) {
                        itemsIndexed(importItems) { index, item ->
                            Row(
                                modifier = Modifier.fillMaxWidth().padding(10.dp),
                                verticalAlignment = Alignment.CenterVertically
                            ) {
                                Text(item["title"].orEmpty(), Modifier.weight(1f))
                                Switch(
                                    checked = selectedItems[index] != false,
                                    enabled = item["category"] != "Failed!",
                                    onCheckedChange = { selectedItems[index] = it }
                                )
                            }
                        }
                    }
                    Row(Modifier.fillMaxWidth().padding(10.dp)) {
                        Button(onClick = { finish() }, modifier = Modifier.weight(1f)) {
                            Text(stringResource(R.string.cancel))
                        }
                        Button(
                            onClick = {
                                BackupUtils.importContents(
                                    applicationContext,
                                    this@BackupImportChooserActivity,
                                    getFinalImportObject()
                                )
                                ToastUtils.showToast(this@BackupImportChooserActivity, R.string.finish)
                                finish()
                            },
                            modifier = Modifier.weight(1f)
                        ) {
                            Text(stringResource(R.string.finish))
                        }
                    }
                }
            }
        }
    }

    private fun getFinalImportObject(): JSONObject {
        val jsonObject = importObject ?: return JSONObject()
        importItems.forEachIndexed { index, itemData ->
            if (selectedItems[index] != false) return@forEachIndexed
            val category = itemData["category"] ?: return@forEachIndexed
            val key = itemData["spKey"] ?: return@forEachIndexed
            val array = jsonObject.optJSONArray(category) ?: return@forEachIndexed
            when (category) {
                "generalSettings_boolean", "generalSettings_string",
                "generalSettings_int", "oneKeyList" -> array.optJSONObject(0)?.remove(key)
                "userTimeScheduledTasks", "userTriggerScheduledTasks",
                "userDefinedCategories" -> {
                    for (arrayIndex in 0 until array.length()) {
                        val value = array.optJSONObject(arrayIndex) ?: continue
                        if (key == value.optString("i", "-1")) value.put("doNotImport", true)
                    }
                }
                "uriAutoAllowPkgs_allows", "installPkgs_autoAllowPkgs_allows" -> {
                    jsonObject.remove(category)
                }
            }
        }
        return jsonObject
    }

    private fun generateKeyToStringIdValuePair() {
        // Int 开始
        keyToStringIdValuePair["onClickFunctionStatus"] = getString(R.string.onClickFunctionStatus)
        keyToStringIdValuePair["sortMethodStatus"] = getString(R.string.sortMethodStatus)
        // Int 结束
        // 一键冻结、一键解冻、离开冻结列表 开始
        keyToStringIdValuePair["okff"] = getString(R.string.oneKeyFreezeList)
        keyToStringIdValuePair["okuf"] = getString(R.string.oneKeyUFList)
        keyToStringIdValuePair["foq"] = getString(R.string.freezeOnceQuitList)
        // 一键冻结、一键解冻、离开冻结列表 结束
        // 安装应用请求、URI 请求白名单 开始
        keyToStringIdValuePair["uriAutoAllowPkgs_allows"] = getString(R.string.uriAutoAllowList)
        keyToStringIdValuePair["installPkgs_autoAllowPkgs_allows"] =
            getString(R.string.ipaAutoAllowList)
        // 安装应用请求、URI 请求白名单 结束
    }

    private fun generateList(
        jsonObject: JSONObject,
        list: ArrayList<MutableMap<String, String>>
    ) {
        val jsonKeysIterator = jsonObject.keys()
        while (jsonKeysIterator.hasNext()) {
            when (jsonKeysIterator.next()) {
                "generalSettings_boolean" -> generateGeneralSettingsBooleanList(jsonObject, list)
                "generalSettings_string" -> generateGeneralSettingsStringList(jsonObject, list)
                "generalSettings_int" -> generateGeneralSettingsIntList(jsonObject, list)
                "oneKeyList" -> generateOneKeyList(jsonObject, list)
                "userTimeScheduledTasks" -> generateUserTimeScheduledTasksList(jsonObject, list)
                "userTriggerScheduledTasks" -> generateUserTriggerScheduledTasksList(
                    jsonObject,
                    list
                )
                "userDefinedCategories" -> generateUserDefinedCategoriesList(jsonObject, list)
                "uriAutoAllowPkgs_allows" -> generateUriAutoAllowPkgsList(jsonObject, list)
                "installPkgs_autoAllowPkgs_allows" -> generateInstallPkgsAutoAllowPkgsList(
                    jsonObject,
                    list
                )
                else -> {}
            }
        }
    }

    private fun generateGeneralSettingsBooleanList(
        jsonObject: JSONObject,
        list: ArrayList<MutableMap<String, String>>
    ) {
        val array = jsonObject.optJSONArray("generalSettings_boolean") ?: return
        val generalSettingsBooleanJSONObject = array.optJSONObject(0) ?: return
        val it = generalSettingsBooleanJSONObject.keys()
        val moreSettingsDashLineLabel = getString(R.string.moreSettingsDashLineLabel)
        var s: String?
        while (it.hasNext()) {
            s = it.next() ?: continue
            if (DefaultSharedPreferenceStorageBooleanKeys.firstIconEnabled.name == s || DefaultSharedPreferenceStorageBooleanKeys.secondIconEnabled.name == s || DefaultSharedPreferenceStorageBooleanKeys.thirdIconEnabled.name == s || DefaultMultiProcessMMKVStorageBooleanKeys.enableAuthentication.name == s) {
                continue
            }
            var key: cf.playhi.freezeyou.storage.key.AbstractKey<Boolean>? = null
            try {
                key = DefaultSharedPreferenceStorageBooleanKeys.valueOf(s)
            } catch (ignored: IllegalArgumentException) {
            }
            if (key == null) {
                try {
                    key = DefaultMultiProcessMMKVStorageBooleanKeys.valueOf(s)
                } catch (ignored: IllegalArgumentException) {
                }
            }
            if (key == null) continue
            val keyValuePair = HashMap<String, String>()
            keyValuePair["title"] = String.format(moreSettingsDashLineLabel, getString(key.titleTextStringId()))
            keyValuePair["spKey"] = s
            keyValuePair["category"] = "generalSettings_boolean"
            list.add(keyValuePair)
        }
    }

    private fun generateGeneralSettingsIntList(
        jsonObject: JSONObject,
        list: ArrayList<MutableMap<String, String>>
    ) {
        val array = jsonObject.optJSONArray("generalSettings_int") ?: return
        val generalSettingsIntJSONObject = array.optJSONObject(0) ?: return
        val it = generalSettingsIntJSONObject.keys()
        var s: String
        while (it.hasNext()) {
            s = it.next()
            when (s) {
                "onClickFunctionStatus", "sortMethodStatus" -> {
                    val keyValuePair = HashMap<String, String>()
                    keyValuePair["title"] =
                        if (keyToStringIdValuePair.containsKey(s)) keyToStringIdValuePair[s]!! else s
                    keyValuePair["spKey"] = s
                    keyValuePair["category"] = "generalSettings_int"
                    list.add(keyValuePair)
                }
                else -> {}
            }
        }
    }

    private fun generateGeneralSettingsStringList(
        jsonObject: JSONObject,
        list: ArrayList<MutableMap<String, String>>
    ) {
        val array = jsonObject.optJSONArray("generalSettings_string") ?: return
        val generalSettingsStringJSONObject = array.optJSONObject(0) ?: return
        val moreSettingsDashLineLabel = getString(R.string.moreSettingsDashLineLabel)
        val it = generalSettingsStringJSONObject.keys()
        var s: String?
        while (it.hasNext()) {
            s = it.next() ?: continue
            var key: cf.playhi.freezeyou.storage.key.AbstractKey<String?>? = null
            try {
                key = DefaultSharedPreferenceStorageStringKeys.valueOf(s)
            } catch (ignored: IllegalArgumentException) {
            }
            if (key == null) {
                try {
                    key = DefaultMultiProcessMMKVStorageStringKeys.valueOf(s)
                } catch (ignored: IllegalArgumentException) {
                }
            }
            if (key == null) continue
            val keyValuePair = HashMap<String, String>()
            keyValuePair["title"] = String.format(moreSettingsDashLineLabel, getString(key.titleTextStringId()))
            keyValuePair["spKey"] = s
            keyValuePair["category"] = "generalSettings_string"
            list.add(keyValuePair)
        }
    }

    private fun generateOneKeyList(
        jsonObject: JSONObject,
        list: ArrayList<MutableMap<String, String>>
    ) {
        val array = jsonObject.optJSONArray("oneKeyList") ?: return
        val generalOneKeyJSONObject = array.optJSONObject(0) ?: return
        val it = generalOneKeyJSONObject.keys()
        var s: String
        while (it.hasNext()) {
            s = it.next()
            when (s) {
                "okff", "okuf", "foq" -> {
                    val keyValuePair = HashMap<String, String>()
                    keyValuePair["title"] =
                        if (keyToStringIdValuePair.containsKey(s)) keyToStringIdValuePair[s]!! else s
                    keyValuePair["spKey"] = s
                    keyValuePair["category"] = "oneKeyList"
                    list.add(keyValuePair)
                }
                else -> {}
            }
        }
    }

    private fun generateUserTimeScheduledTasksList(
        jsonObject: JSONObject,
        list: ArrayList<MutableMap<String, String>>
    ) {
        val array = jsonObject.optJSONArray("userTimeScheduledTasks") ?: return
        var oneUserTimeScheduledTaskJSONObject: JSONObject?
        val scheduledTaskDashLineLabel = getString(R.string.scheduledTaskDashLineLabel)
        val defaultLabel = getString(R.string.label)
        for (i in 0 until array.length()) {
            oneUserTimeScheduledTaskJSONObject = array.optJSONObject(i)
            if (oneUserTimeScheduledTaskJSONObject == null) {
                continue
            }
            val keyValuePair = HashMap<String, String>()
            keyValuePair["title"] = String.format(
                scheduledTaskDashLineLabel,
                oneUserTimeScheduledTaskJSONObject.optString("label", defaultLabel)
            )
            try {
                oneUserTimeScheduledTaskJSONObject.put("i", i.toString())
            } catch (e: JSONException) {
                e.printStackTrace()
            }
            keyValuePair["spKey"] = i.toString()
            keyValuePair["category"] = "userTimeScheduledTasks"
            list.add(keyValuePair)
        }
    }

    private fun generateUserTriggerScheduledTasksList(
        jsonObject: JSONObject,
        list: ArrayList<MutableMap<String, String>>
    ) {
        val array = jsonObject.optJSONArray("userTriggerScheduledTasks") ?: return
        var oneUserTriggerScheduledTaskJSONObject: JSONObject?
        val scheduledTaskDashLineLabel = getString(R.string.scheduledTaskDashLineLabel)
        val defaultLabel = getString(R.string.label)
        for (i in 0 until array.length()) {
            oneUserTriggerScheduledTaskJSONObject = array.optJSONObject(i)
            if (oneUserTriggerScheduledTaskJSONObject == null) {
                continue
            }
            val keyValuePair = HashMap<String, String>()
            keyValuePair["title"] = String.format(
                scheduledTaskDashLineLabel,
                oneUserTriggerScheduledTaskJSONObject.optString("label", defaultLabel)
            )
            try {
                oneUserTriggerScheduledTaskJSONObject.put("i", i.toString())
            } catch (e: JSONException) {
                e.printStackTrace()
            }
            keyValuePair["spKey"] = i.toString()
            keyValuePair["category"] = "userTriggerScheduledTasks"
            list.add(keyValuePair)
        }
    }

    private fun generateUserDefinedCategoriesList(
        jsonObject: JSONObject,
        list: ArrayList<MutableMap<String, String>>
    ) {
        val array = jsonObject.optJSONArray("userDefinedCategories") ?: return
        var oneUserDefinedCategoriesJSONObject: JSONObject?
        val myCustomizationDashLineLabel = getString(R.string.myCustomizationDashLineLabel)
        val defaultLabel = ""
        for (i in 0 until array.length()) {
            oneUserDefinedCategoriesJSONObject = array.optJSONObject(i)
            if (oneUserDefinedCategoriesJSONObject == null) {
                continue
            }
            val keyValuePair = HashMap<String, String>()
            keyValuePair["title"] = String.format(
                myCustomizationDashLineLabel,
                String(
                    Base64.decode(
                        oneUserDefinedCategoriesJSONObject.optString(
                            "label", defaultLabel
                        ),
                        Base64.DEFAULT
                    )
                )
            )
            try {
                oneUserDefinedCategoriesJSONObject.put("i", i.toString())
            } catch (e: JSONException) {
                e.printStackTrace()
            }
            keyValuePair["spKey"] = i.toString()
            keyValuePair["category"] = "userDefinedCategories"
            list.add(keyValuePair)
        }
    }

    private fun generateUriAutoAllowPkgsList(
        jsonObject: JSONObject,
        list: ArrayList<MutableMap<String, String>>
    ) {
        val array = jsonObject.optJSONArray("uriAutoAllowPkgs_allows") ?: return
        val jObj = array.optJSONObject(0) ?: return
        val keyValuePair = HashMap<String, String>()
        keyValuePair["title"] = if (keyToStringIdValuePair.containsKey("uriAutoAllowPkgs_allows"))
            keyToStringIdValuePair["uriAutoAllowPkgs_allows"]!!
        else
            "uriAutoAllowPkgs_allows"
        keyValuePair["spKey"] = "uriAutoAllowPkgs_allows"
        keyValuePair["category"] = "uriAutoAllowPkgs_allows"
        list.add(keyValuePair)
    }

    private fun generateInstallPkgsAutoAllowPkgsList(
        jsonObject: JSONObject,
        list: ArrayList<MutableMap<String, String>>
    ) {
        val array = jsonObject.optJSONArray("installPkgs_autoAllowPkgs_allows") ?: return
        val jObj = array.optJSONObject(0) ?: return
        val keyValuePair = HashMap<String, String>()
        keyValuePair["title"] =
            if (keyToStringIdValuePair.containsKey("installPkgs_autoAllowPkgs_allows"))
                keyToStringIdValuePair["installPkgs_autoAllowPkgs_allows"]!!
            else
                "installPkgs_autoAllowPkgs_allows"
        keyValuePair["spKey"] = "installPkgs_autoAllowPkgs_allows"
        keyValuePair["category"] = "installPkgs_autoAllowPkgs_allows"
        list.add(keyValuePair)
    }
}
